### DISCUSSION TITLE
Installing dependencies with pnpm instead of git submodules

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#508] [Ideas]
See my discussion with Drake Evans on Twitter:

https://twitter.com/DrakeEvansV1/status/1659612843849793573

**Comment by andreivladbrg** [comment]
That's interesting. 

How would the remappings and the recursive submodules work? For example in our `v2-periphery` we import the prb-math types from `v2-core/src/types/Math.sol` and if we don't install the dependencies recursively we would be unable to utilize them.

Is there going to affect in any way the installation speed?

**Reply by PaulRBerg** [reply]
Remappings are easy to solve - we just define `allow_paths` in `foundry.toml`, see this example:

https://github.com/clabby/example-pnpm-forge-monorepo/blob/dd2a21d0545f7818597a6b4a7bb5611386153dae/packages/contracts/foundry.toml

> import the `prb-math` types from `v2-core/src/types/Math.sol`

PRBMath is a [peer dependency](https://github.com/sablier-labs/v2-core/blob/70cd2f38aae65501ca95cf1e0dbf496001c3d3a1/package.json#L44) of V2 Core so users get a copy of the [`@prb/math`](https://www.npmjs.com/package/@prb/math) package automatically.

> Is there going to affect in any way the installation speed?

Installation times would be faster, actually, because Pnpm hoists dependencies (i.e. it does not install them multiple times, as git submodules do).

**Comment by PaulRBerg** [comment]
@andreivladbrg could you explore the possibility of ditching submodules in favor of Node.js dependencies?

**Reply by andreivladbrg** [reply]
I worked on this today (see [here](https://github.com/sablier-labs/v2-core/tree/build/modules-to-npm)), it seems to be possible and in might be a good choice to switch to pnpm, at least in terms of time spent installing dependencies:

<img width="600" alt="image" src="https://github.com/sablier-labs/v2-core/assets/99738872/24d707ed-37d0-468a-a48c-44b8cda9ec93">

We have to test it in practice to see how would it work for integrators installing our repo. Will see how it goes on `v2-periphery`

Also, `prb-math` needs to be updated, because of the following [files](https://github.com/PaulRBerg/prb-math/tree/77fa88eda4a4a91b3f3e9431df291292c26b6c71/src/test). They must be updated to hard-coded `test-utils` in `src`, instead of importing them from `../../test/utils/` path, `test` is not published in the package

**Reply by PaulRBerg** [reply]
For future reference: we ended up including the `test/utils` directory in the npm package:

https://github.com/PaulRBerg/prb-math/pull/211

> `test` is not published in the package

`test/utils` will be published in v4.0.2

**Comment by andreivladbrg** [comment]
### Context 

Using git submodules has started to be a pain in the ass because of the installation times.

I've been working on this, and it's a more complex idea than expected.

The installation process for `@sablier/v2-core`, both for the project itself and for external projects installing it, would be much faster. However, there are some edge cases.

External projects can be either __foundry__ or __hardhat__ based. I am not considering brownie here.

### Foundry

If the external project is foundry-based, it can be installed via:
1. `yarn add @sablier/v2-core` 
2. `pnpm add @sablier/v2-core` 
3. `forge install sablier-labs/v2-periphery` 

#### 

Any of these versions would require the project to add the following to their remappings file:

```txt
@openzeppelin/contracts/=node_modules/@openzeppelin/contracts/
@prb/math/=node_modules/@prb/math/
```

#### Third party dependencies

By third-party dependencies, I mean the dependencies of the package that's being installed.

For 1. no additional installation is required because it will install its [dependencies](https://github.com/sablier-labs/v2-core/blob/b24966f859b7f6611bf51e8c86ad7547a781c667/package.json#L13-L19) in `node_modules`.

The json file would look like this:
```json
  "dependencies": {
    "@sablier/v2-core": "x.y.z",
   }
```

For 2. and 3. they would also need to manually install `@prb/math/` and `@openzeppelin/contracts/`. For 2, it's necessary because of the way pnpm works with its [symlinks](https://pnpm.io/symlinked-node-modules-structure). The third-party dependencies are placed in `node_modules/.pnpm`.

The json file would look like this:
```json
  "dependencies": {
    "@openzeppelin/contracts": "x.y.z",
    "@prb/math": "x.y.z",
    "@sablier/v2-core": "x.y.z",
   }
```

Import and usage:
```sol
import { ISablierV2Lockup } from "@sablier/v2-core/src/interfaces/ISablierV2Lockup.sol";
import { Broker, LockupDynamic, LockupLinear } from "@sablier/v2-core/src/types/DataTypes.sol";
```

### Hardhat

If the external project is hardhat-based, it can be installed via:
1. `yarn add @sablier/v2-core` 
2. `pnpm add @sablier/v2-core` 

No other remappings configuration is required.

#### Third-party dependencies

The same as foundry's 1 and 2.

Import and usage:
```sol
import { ISablierV2Lockup } from "@sablier/v2-core/src/interfaces/ISablierV2Lockup.sol";
import { Broker, LockupDynamic, LockupLinear } from "@sablier/v2-core/src/types/DataTypes.sol";
```


### Question

Should we "force" our integrators that use `foundry`, which already have installed openzeppelin to add a new remapping to their file? IMO yes, we should recommend NPM installations in favor of git submodules and we should make a best practice for our remappings style.

Given the requirement of manually installing third-party dependencies for `pnpm`, should we switch to `yarn`? I'm aware that it's slightly slower.

### Conclusions

In case I am not missing something, it's possible to completely switch to NPM packages from git submodules, and in my opinion, it looks like the correct decision.

What do you think?

**Reply by andreivladbrg** [reply]
**Relevant note**: currently it takes 1m 43s to run  [git submodule --init --recursive](https://app.warp.dev/block/AHX3f40hJLsoSEbqkRMOdK) on `v2-periphery`

![image](https://github.com/sablier-labs/v2-core/assets/99738872/ba3f8b7c-8873-4dfc-b7d3-d6fc01719743)

**Reply by andreivladbrg** [reply]
I have to mention that it is not 100% necessary to manually install openzeppelin and prb math when using `pnpm`, if you don't, the remappings should look like this:

```txt
@openzeppelin/contracts/=.pnpm/@openzeppelin+contracts@x.y.z/node_modules/@openzeppelin/contracts/
@prb/math/=node_modules/.pnpm/@prb+math@x.y.z/node_modules/@prb/math/
```

We should not recommend this, given how ugly it is.

**Reply by andreivladbrg** [reply]
I realized that these:

>>Any of these versions would require the project to add the following to their remappings file:
>>```txt
>>@openzeppelin/contracts/=node_modules/@openzeppelin/contracts/
>>@prb/math/=node_modules/@prb/math/
>>```

Can be fixed if we include the `remappings.txt` file in the NPM package. 

Would it be a good choice to publish them? Tbh, I am not sure, as it may collide with the project installing the `v2-core`  as a dependency. Maybe we should have a dedicated remappings file to publish (which would include only openzeppelin and prb-math) and one for the project.

**Reply by PaulRBerg** [reply]
Thanks for all this @andreivladbrg - just letting you know that I will review this later. Hopefully during the weekend!

**Reply by PaulRBerg** [reply]
> For 2, it's necessary because of the way pnpm works with its [symlinks](https://pnpm.io/symlinked-node-modules-structure).

Surely there must be a way to force Pnpm to not put the dependencies under `node_modules/.pnpm` without asking users to manually install those packages?

> Can be fixed if we include the remappings.txt file in the NPM package.

Correct.

> Should we "force" our integrators that use foundry

We cannot force people to do anything. But we can present Foundry as the recommended approach.

> Given the requirement of manually installing third-party dependencies for pnpm, should we switch to yarn?

Again, we cannot force users to switch to anything. But we can definitely recommend Yarn over Pnpm (in our docs) if you believe this would be easier to handle.

> Would it be a good choice to publish them?

Yes, definitely. Thanks to this PR:

https://github.com/foundry-rs/foundry/pull/5532

The remappings will be de-duped.

> Maybe we should have a dedicated remappings file to publish (which would include only openzeppelin and prb-math) and one for the project.

What's wrong with including all remappings?

> In case I am not missing something, it's possible to completely switch to NPM packages from git submodules ... it looks like the correct decision.
>
> What do you think?

Agree. Thanks for your detailed analysis.

**Reply by PaulRBerg** [reply]
@andreivladbrg waiting for your responses here.

**Reply by andreivladbrg** [reply]
not sure what to responde here, i've created the PR https://github.com/sablier-labs/v2-core/pull/734

**Reply by PaulRBerg** [reply]
Ok I will review there

**Reply by PaulRBerg** [reply]
@andreivladbrg can you please answer my question above:

> Surely there must be a way to force Pnpm to not put the dependencies under `node_modules/.pnpm` without asking users to manually install those packages?

For maximum clarity, I am referring to integrators, not to us.

**Reply by PaulRBerg** [reply]
It would be worth asking the Pnpm maintainers for help - telling them that we would like to continue using Pnpm instead of Yarn:

https://github.com/orgs/pnpm/discussions

**Comment by PaulRBerg** [comment]
Closing as we have implemented this proposal in https://github.com/sablier-labs/v2-core/pull/734.
